src/lib/FlatMemoryInterface.cc

```diff
@@ -5,40 +5,38 @@
 
 namespace simeng {
 
-FlatMemoryInterface::FlatMemoryInterface(
-    std::shared_ptr<simeng::memory::Mem> memory) {
-  memory_ = memory;
-  size_ = memory_->getMemorySize();
+FlatMemoryInterface::FlatMemoryInterface(std::shared_ptr<memory::MMU> mmu,
+                                         size_t memSize) {
+  mmu_ = mmu;
+  size_ = memSize;
 }
 
 void FlatMemoryInterface::requestRead(const MemoryAccessTarget& target,
                                       uint64_t requestId) {
-  if (target.address + target.size > size_) {
-    // Read outside of memory; return an invalid value to signal a fault
-    completedReads_.push_back({target, RegisterValue(), requestId});
-    return;
-  }
-
-  simeng::memory::ReadRespPacket* resp =
-      (simeng::memory::ReadRespPacket*)memory_->requestAccess(
-          new simeng::memory::ReadPacket(target.address, target.size));
-
-  // Copy the data at the requested memory address into a RegisterValue
-  completedReads_.push_back(
-      {target, RegisterValue(resp->data, resp->bytesRead), requestId});
-  delete resp;
+  auto fn = [&, this](memory::DataPacket* dpkt = NULL) -> void {
+    if (dpkt == NULL) {
+      this->completedReads_.push_back({target, RegisterValue(), requestId});
+      return;
+    }
+    memory::ReadRespPacket* resp = (memory::ReadRespPacket*)dpkt;
+    this->completedReads_.push_back(
+        {target, RegisterValue(resp->data, resp->bytesRead), requestId});
+    delete resp;
+  };
+
+  mmu_->bufferRequest(new memory::ReadPacket(target.address, target.size), fn);
 }
 
 void FlatMemoryInterface::requestWrite(const MemoryAccessTarget& target,
                                        const RegisterValue& data) {
-  assert(target.address + target.size <= size_ &&
-         "Attempted to write beyond memory limit");
-
-  simeng::memory::WriteRespPacket* resp =
-      (simeng::memory::WriteRespPacket*)memory_->requestAccess(
-          new simeng::memory::WritePacket(target.address, target.size,
-                                          data.getAsVector<char>()));
-  delete resp;
+  auto fn = [&](memory::DataPacket* dpkt = NULL) -> void {
+    if (dpkt == NULL) return;
```

Comment:
done.

Reply:
In case the diff doesn't work here is the new comment:
"Sending an empty RegisterValue here signifies a data abort exception as we haven't recieved a response from memory"